### DISCUSSION TITLE
How to plot three axes for LineSeries?

### DISCUSSION THREAD

**Discussion by redundancer** [oxyplot/oxyplot#2003] [Q&A]
I can easily add two axes to a LineSeries. The first one I put at the bottom and it automatically scales with the data's X-value. The second one I put at the left and it automatically scales with the Y-Value. Now I want a third axis which scales with a custom value. For example I define my data as

```csharp
public class CustomDataPoint : IDataPointProvider
        {
            public double X { get; }
            public double Y { get; }
            public double Z => X * Y;

            public CustomDataPoint(double x, double y)
            {
                X = x;
                Y = y;
            }

            public DataPoint GetDataPoint()
            {
                return new DataPoint(X, Y);
            }
        }
```

then I would like the thrid axes to be at the top and scale with the Z-value. How can I do so?

**Comment by redundancer** [comment]
Can anyone help please?

**Comment by VisualMelon** [comment]
I don't understand what you're asking for: can produce a mock-up drawing to explain illustrate the result you want?

**Reply by redundancer** [reply]
Sure! Here is some example ViewModel. I use [Prism](https://prismlibrary.com/docs/wpf/legacy/Introduction.html) to make the code a bit more lightweight:

```csharp
using System;
using System.Linq;
using MathNet.Numerics;
using OxyPlot;
using OxyPlot.Axes;
using OxyPlot.Series;
using Prism.Mvvm;

namespace TmpWpf.ViewModels
{
    public class MainWindowViewModel : BindableBase
    {
        private string _title = "3 Axes Example WPF";
        public string Title
        {
            get { return _title; }
            set { SetProperty(ref _title, value); }
        }

        private PlotModel _plotModel;
        public PlotModel PlotModel
        {
            get { return _plotModel; }
            set { SetProperty(ref _plotModel, value); }
        }

        public MainWindowViewModel()
        {
            CreatePlot();
        }

        private void CreatePlot()
        {
            var range = Generate.LinearRange(0.0, 0.01, 1.2);
            var itemsSource = range.Select(x => new CustomDataPoint(x, F(x)));
            PlotModel = new PlotModel {Title = "Linear series with 3 axes"};
            PlotModel.Series.Add(new LineSeries
            {
                Title = "Line series",
                ItemsSource = itemsSource,
                TrackerFormatString = "{0}\n{1}:\t{2}\n{3}:\t{4}\nZ:\t{Z}"
            });

            PlotModel.Axes.Add(new LinearAxis{Position = AxisPosition.Bottom}); // Scales with X.
            PlotModel.Axes.Add(new LinearAxis{Position = AxisPosition.Left}); // Scales with Y.
            PlotModel.Axes.Add(new LinearAxis{Position = AxisPosition.Top}); // Should scale with Z, but doesn't.
            PlotModel.InvalidatePlot(true);
        }

        private double F(double x)
        {
            return Math.Pow(x, 4) - Math.Pow(x, 3);
        }
    }

    class CustomDataPoint : IDataPointProvider
    {
        public double X { get; }
        public double Y { get; }
        public double Z => X * Y;

        public CustomDataPoint(double x, double y)
        {
            X = x;
            Y = y;
        }

        public DataPoint GetDataPoint()
        {
            return new DataPoint(X, Y);
        }
    }
}
```

My View:
```xaml
<Window x:Class="TmpWpf.Views.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:prism="http://prismlibrary.com/"
        xmlns:oxy="http://oxyplot.org/wpf"
        prism:ViewModelLocator.AutoWireViewModel="True"
        Title="{Binding Title}" Height="350" Width="525">
    <Grid>
        <oxy:PlotView Model="{Binding PlotModel}"/>
    </Grid>
</Window>
```
And this is how the plot looks like:

![grafik](https://github.com/oxyplot/oxyplot/assets/81110847/9f3f81c5-d6a5-49cc-9ad1-e2e776dd1121)

The third axis at the top should display the values of the property Z and when I zoom in it should scale dependent on Z. Is this somehow possible?

One solution would be to have two plots: 1. X as X-Axis and 2. Z as X-Axis, but I would like to display both in one plot.

Thank you!

**Reply by VisualMelon** [reply]
There's nothing in the library which will help you with that, I'm afraid; none of the axes respond to the series beyond determining their bounds. I'm also not sure how you'd generalise the example in the original post: an axes is not suited to showing an arbitrary function: the Z values do not have a nice progression, and there's no guarantee they would even be monotonic

That said, I may be misunderstanding completely: if you just need a way to keep two axes 'in sync', there is discussion and an example in https://github.com/oxyplot/oxyplot/issues/1629 : basically, listen for changes in one axes and update the other manually.

**Reply by redundancer** [reply]
I think the link to #1629  might help, thank you! I would like to explain again. A practical example would be in radiation therapy. The left axis shows the relative volume of a structure. The top axis shows the absolute dose, how much radiation a voxel (or histogram bin) in total gets.  And the bottom axis shows the relative dose, relative to a predefined target dose:

![grafik](https://github.com/oxyplot/oxyplot/assets/81110847/b86e65fb-04a1-4ee2-8274-de7cd40fea48)

Source: https://www.researchgate.net/figure/The-cumulative-dose-volume-histogram-for-PTV-with-and-without-the-applicator-volume-by_fig2_234135460

Now you can read: 80% of the Structure "PTV" gets around 8000cGy radiation (top axis), which is 140% relative to the target dose (bottom axis).

**Reply by VisualMelon** [reply]
That sounds like a nice linear relationship, so should be fine to set up a second linear axis as in #1629: just need to set the bounds of the slave axis as a function of the master axis (and vice versa if it makes sense). Sorry if I got completely the wrong idea from the OP.
